feat: implement GroupNoble

sublimator opened the pull request:
Continuing on with #26 
Also deals with #30 

#### Scripts 

| Test Type | All Groups | Specific Group (For performance while testing) |
|-----------|---------------|-------------|
| All Tests | `npm test` | `GROUP_CONS=noble npm test` |
| ESM Tests | `npm run test:esm` | `GROUP_CONS=noble npm run test:esm` |
| CJS Tests | `npm run test:cjs` | `GROUP_CONS=noble npm run test:cjs` |

#### Notes 

* The Oprf.Group member is [re]configured by a describeGroupTests helper that takes a `declare` function and uses describe.each 
  * The library itself does not honour the env variable.
* @noble/hashes and @noble/curves are listed as "optionalDependencies" (also "devDependencies" for tests) as not strictly required to use the library
* Use Group.fromID factory instead of Group `new` constructor (eye to short vs [ristretto](https://github.com/paulmillr/noble-curves/blob/cf17f7fe01a262396db1b5efdeaff1b5fb31d84a/src/ed25519.ts#L361)/[decaf448](https://github.com/paulmillr/noble-curves/blob/cf17f7fe01a262396db1b5efdeaff1b5fb31d84a/src/ed448.ts#L326C1-L326C1) etc)

sublimator commented on src/index.ts:
```diff
@@ -4,9 +4,12 @@
 import { Oprf } from './oprf.js'
 
 Oprf.Group = GroupConsSjcl
+Oprf.Group = GroupConsNoble
```
TODO obviously

sublimator replied:
see: https://github.com/cloudflare/voprf-ts/pull/38#issuecomment-1605261006

sublimator commented on test/server.test.ts:
```diff
@@ -63,6 +63,8 @@ describe.each(Object.entries(Oprf.Suite))('supportsWebCrypto', (name, id) => {
-        expect(ev0).toEqual(ev1)
+        // serialize first, as bigint can't be serialized by jest-worker
+        // see: https://github.com/jestjs/jest/issues/11617
+        expect(ev0.serialize()).toEqual(ev1.serialize())
```
TODO

sublimator replied:
JSON.stringify instead with a BigInt handler ??

sublimator replied:
k, using Evaluation#isEqual

sublimator commented on package.json:
```diff
@@ -29,19 +31,19 @@
-        "typescript": "4.8.4",
-        "sjcl": "1.0.8"
+        "eslint-plugin-security": "1.7.1",
+        "jest": "^29.5.0",
```
TODO: pin

armfazh commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    serialize(compressed = true): Uint8Array {
+        if (this.p.equals(getCurve(this.g.id).Point.ZERO)) {
```
```suggestion
        if (this.p.isIdentity()) {
```

armfazh commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,338 @@
+}
+const curves: Record<GroupID, CurveDef> = {
+    [GroupNb.ID.P256]: {
+        fp: FieldCons(p256.CURVE.n),
```
```suggestion
        fp: p256.CURVE.Fp,
```

sublimator replied:
Actually, that's Fp, the prime, not the order as is used here. I've renamed it to `fn`

sublimator replied:
n is of course prime, but I guess it makes a a bit clearer

armfazh commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,338 @@
+    static hash(g: GroupNb, msg: Uint8Array, dst: Uint8Array): ScalarNb {
+        const { hash, L } = getHashParams(g.id)
+        const s = expand_message_xmd(msg, dst, L, hash)
+        return new ScalarNb(g, bytesToNumberBE(s))
```
here there is no reduction modulo group order, or better invoke `hash_to_field`

sublimator replied:
added a modulo by group order

sublimator replied:
> better invoke hash_to_field

btw, I had a quick look into that, but wasn't able to get the tests to pass. 
Will look again later

sublimator replied:
```typescript
    static async hash(g: GroupSj, msg: Uint8Array, dst: Uint8Array): Promise<ScalarSj> {
        const { hash, L } = getHashParams(g.id)
        const bytes = await expandXMD(hash, msg, dst, L)
        const array = Array.from(bytes)
        const bitArr = sjcl.codec.bytes.toBits(array)
        const k = sjcl.bn.fromBits(bitArr).mod(getCurve(g.id).r)
        return new ScalarSj(g, k)
    }
```

Actually, the sjcl implementation simply uses expandMD ?

sublimator replied:
@paulmillr @armfazh

Ok, I updated it to use hash_to_field

I noticed something that I need to understand better. 
hashToCurve -> element uses hash_to_field with p set (by @noble, see [here](https://github.com/paulmillr/noble-curves/blob/c2edc978683c1a688eecc0f020a80a1ff8ded749/src/p256.ts#L11-L41)) to the curve's prime, whereas we seem to need to use the sub group order as p for the tests to pass here.

sublimator replied:
Wait, p is the size of the field of possible values for x,y in an element ?
Whereas n is size of of the field of possible elements, thus suitable for scalar multiplication  ?

armfazh replied:
In an elliptic curve there are two important primes:
p, the prime used to define the curve params and the (x,y,z) coordinates.
n, (or r), the prime that defines the group order. Equiv, defines the range for scalars.

Since both are primes, you can always define a prime field (GF) with each of them. 

For hash_to_curve, we use internally `hash_to_field( prime = p)` (i.e., setting the prime to p) for getting candidates of coordinates.

However, for random_scalar, we can use internally `hash_to_field( prime = n )` (i.e., setting the prime to n), so we can get scalars in such a range.

Hope this clarifies.

sublimator replied:
Thanks! Yeah, that makes sense

armfazh commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,338 @@
+    static hash(g: GroupNb, msg: Uint8Array, dst: Uint8Array): EltNb {
+        const h2c = getCurve(g.id).h2c
+        const DST = new TextDecoder().decode(dst)
+        const p = h2c(msg, { DST }) as ProjPointType<bigint>
```
hash to curve accepts `Uint8Array`, no need to use `TextDecoder`

sublimator replied:
Not sure if a bug, but it seemed to throw saying it required a string

sublimator replied:
https://github.com/paulmillr/noble-curves/issues/57

sublimator replied:
Fixed by https://github.com/paulmillr/noble-curves/pull/58

sublimator commented:
Thanks!

Will sort when on keyboard

armfazh commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,328 @@
+    GroupIDs,
+    Scalar,
+    ScalarCons
+} from './groupTypes'
```
```suggestion
} from './groupTypes.js'
```

sublimator commented:
@armfazh 

I added some tsconfig.cjs.json files, and added them as a reference to the root tsconfig.json. 
However it slows things down considerably during development.

I also note that this is declaring "type": "module" in the package.json, which, due to node module resolution means any consumers will have to be using ESM, regardless of any "exports" configuration:

```

(base) ➜  voprf-ts git:(nd-wip-2023-06-20) ✗ node
Welcome to Node.js v18.16.0.
Type ".help" for more information.
> require('.')
Uncaught:
Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/nicholasdudfield/projects/voprf-ts/lib/cjs/src/index.js not supported.
index.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /Users/nicholasdudfield/projects/voprf-ts/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

    at __node_internal_captureLargerStackTrace (node:internal/errors:490:5)
    at new NodeError (node:internal/errors:399:5)
    at Module._extensions..js (node:internal/modules/cjs/loader:1281:19)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18) {
  code: 'ERR_REQUIRE_ESM'
```

I tried neutralising the package.json by removing "type":"module", however jest started failing as it didn't know to interpret .js files as ESM:

```
    Details:

    /Users/nicholasdudfield/projects/voprf-ts/lib/test/vectors.test.js:5
    import { derivePrivateKey, generatePublicKey, Oprf, OPRFClient, OPRFServer, POPRFClient, POPRFServer, VOPRFClient, VOPRFServer } from '../src/index.js';
    ^^^^^^

    SyntaxError: Cannot use import statement outside a module
```

I will try again later, but if you have any ideas? 

Thinking out loud: 
Perhaps there could be a separate cjs build step to create a cjs package.json (minus "type": "module"), compile all the code (including the test code) also as cjs, and a cjs specific test command ? Otherwise, there needs to be some way of configuring jest to run the tests without "type": "module" ?

sublimator commented:
@armfazh 

Ok, I've got something that seems to work. I removed "type" : "module" from the root package.json and committed lib/{esm,cjs}/package.json files setting their respective "type"s. At the moment the test/build commands simply build and test both cjs and esm at once. SCJL is the default group, and the tests can be configured to test noble by setting the environment variable `GROUP_CONS=noble`

sublimator commented:
@armfazh @lukevalenta 

?

armfazh commented on test/dleq.test.ts:
```diff
@@ -4,11 +4,11 @@
 
 import { DLEQParams, DLEQProof, DLEQProver, Elt, Oprf, Scalar } from '../src/index.js'
-
+import './setGroup.js'
```
Let's use `describe.each([groupNoble, groupSjcl])` to run both versions using jest.
So no need of ENV variables.

sublimator replied:
```
import { GroupConsNoble } from '../src/groupNoble.js'
import { GroupConsSjcl } from '../src/groupSjcl.js'
import { GroupCons, Oprf } from '../src/index.js'

const groupConsMatch = process.env.GROUP_CONS

export const testGroups = (
    [
        ['noble', GroupConsNoble],
        ['sjcl', GroupConsSjcl]
    ] as const
).filter(([name]) => {
    return !groupConsMatch || name === groupConsMatch
})

export function describeGroupTests(declare: (group: GroupCons) => void) {
    describe.each(testGroups)(`Group-%s`, (_groupName, groupCons) => {
        Oprf.Group = groupCons
        declare(groupCons)
    })
}
```

I believe the above (or something close) ought to keep us both happy :) !

armfazh commented on test/dleq.test.ts:
```diff
@@ -4,11 +4,11 @@
+import './setGroup.js'
 import { serdeClass } from './util.js'
 
 describe.each(Object.entries(Oprf.Group.ID))('DLEQ', (groupName, id) => {
```
```suggestion
import { GroupConsNoble } from '../src/groupNoble.js'
import { GroupConsSjcl } from '../src/groupSjcl.js'

describe.each([
    ["noble",GroupConsNoble],
    ["sjcl",GroupConsSjcl]
])(`Group-%s`,(_gropName :string, g :GroupCons)=>{
    Oprf.Group = g
describe.each(Object.entries(Oprf.Group.ID))('DLEQ', (groupName, id) => {
```

sublimator replied:
I think maybe we could import the test declaration from some file that honours an env var.
Or perhaps jest has some inbuilt test filtering ? In any case, it's desirable to not have to sit through the slow sjcl tests

armfazh commented on test/setGroup.ts:
```diff
@@ -0,0 +1,7 @@
+// noinspection ES6PreferShortImport
```
remove this file

armfazh commented on .github/workflows/node.yml:
```diff
@@ -26,4 +26,5 @@ jobs:
       - run: npm run lint
       - run: npm run build
       - run: npm test
+      - run: GROUP_CONS=noble npm test
```
no changes needed if we run variants in Jest.

sublimator replied:
Yeah, I think defaulting to running all the tests in jest makes sense, though equally, a way to test only a single group without having to edit some file, also does.

armfazh reviewed (changes requested):
Hi, here are some suggested changes.

sublimator commented:
I found it helpful to be able to target a specific group for performance
sake. These tests can be quite slow. Maybe there is a middle ground here.

Am currently on a trip(and on mobile), though back soon.

Thanks.

sublimator commented on src/sjcl/index.js:
```diff
@@ -3265,7 +3265,7 @@ sjcl.ecc.basicKey = {
     * @return x and y as bitArrays
     */
-    this.get = function() {
+    this.for = function() {
```
oops!!!

sublimator commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,337 @@
+        return EltNb.gen(this).mul(s)
+    }
+
+    async randomScalar(): Promise<ScalarNb> {
```
Some linters actually enforce consistent use of async, which is something I can get behind, but if that's the convention here, I can go with it. Maybe there's a linting rule this project could set up ?

sublimator replied:
async randomScalar(): Promise<ScalarNb> {
        const msg = crypto.getRandomValues(new Uint8Array(this.size))
        return ScalarNb.hash(this, msg, new Uint8Array())
    }

Actually, looking at this more, I can see that the async modifier is what makes the simple `return value` upgrade to a Promise, rather than having to use Promise.resolve(value) 

Our Group interfaces are async (using Promises), so the async is actually required?

armfazh replied:
until there is a linter rule enabled to get this uniformly, I consider that using `async` only when inside the body of the function there is an `await` keyword.

A function is known to be async iff it returns a `Promise`.

sublimator replied:
> A function is known to be async iff it returns a Promise.

Let's dig into this a bit more. 

The randomScalar method of our Group interface is asynchronous so that it can support the sjcl/webcrypto (used for expandMD) implementation. 

> scalarNb.hash(this, msg, new Uint8Array())

This is a sync operation, and not a promise, it's true, but by putting `async` on the function it "upgrades" the return value to a Promise so it will conform to our interface.

The alternative here is to use Promise.resolve(), which I will do: 

```typescript
    randomScalar(): Promise<ScalarNb> {
        const msg = crypto.getRandomValues(new Uint8Array(this.size))
        return Promise.resolve(ScalarNb.hash(this, msg, new Uint8Array()))
    }
```

sublimator commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,337 @@
+    static readonly Elt: EltCons = EltNb
+    static readonly Scalar: ScalarCons = ScalarNb
+    static readonly ID = GroupIDs
+    static for(gid: GroupID) {
```
Yeah, it's potentially a YAGNI, but it was looking forward to when ristretto/decaf are implemented and it would likely be cleaner to use separate classes. Using new() would mean having to build a single class for all. 

@armfazh 
Does that make sense?

armfazh replied:
not so clear right now, until we have the other groups.
One concern is more about using `for`, can we replace it with `create`, `createFromID` or `fromID`?

sublimator replied:
Ok, will use fromID

sublimator commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,337 @@
+): GroupParams {
+    const curve = Reflect.get(module, k)
+    return {
+        fpN: Field(curve.CURVE.n),
```
Good call

sublimator commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,337 @@
+        compat(this, k2)
+        compat(this, a)
+        const el = this.p.multiplyAndAddUnsafe(a.p, k1.k, k2.k)
+        if (!el) throw new Error('result is zero')
```
```
    /**
     * Efficiently calculate `aP + bQ`. Unsafe, can expose private key, if used incorrectly.
     * Not using Strauss-Shamir trick: precomputation tables are faster.
     * The trick could be useful if both P and Q are not G (not in our case).
     * @returns non-zero affine point
     */
    multiplyAndAddUnsafe(Q: Point, a: bigint, b: bigint): Point | undefined {
      const G = Point.BASE; // No Strauss-Shamir trick: we have 10% faster G precomputes
      const mul = (
        P: Point,
        a: bigint // Select faster multiply() method
      ) => (a === _0n || a === _1n || !P.equals(G) ? P.multiplyUnsafe(a) : P.multiply(a));
      const sum = mul(this, a).add(mul(Q, b));
      return sum.is0() ? undefined : sum;
    }
```

It seems multiplyAndAddUnsafe returns a non zero Point

This one is probably better fielded by @paulmillr

sublimator replied:
For the moment I'm just returning a zero point when multiplyAndAddUnsafe returns undefined (the point was zero) 

Any better ideas ?

armfazh replied:
it seems to me that this need to be solved in the noble package.

sublimator replied:
The workaround should be OK to get things moving though, no?

sublimator commented:
@armfazh 

I've addressed your concerns/comments. I've left the `for` static method for now, in 
anticipation of separate Group/Elt/Scalar classes used by decaf/ristretto. 

Thanks!

armfazh commented on src/groupNoble.ts:
```diff
@@ -0,0 +1,337 @@
+        return ScalarNb.hash(this, msg, dst)
+    }
+
+    eltDes: Deserializer<EltNb> = {
+        size: (compressed?: boolean) => {
+            return this.eltSize(compressed)
+        },
+        deserialize: (b: Uint8Array) => {
+            return this.desElt(b)
+        }
+    }
+    scalarDes: Deserializer<ScalarNb> = {
+        size: () => {
+            return this.scalarSize()
+        },
+        deserialize: (b: Uint8Array) => {
+            return this.desScalar(b)
+        }
+    }
```
bump

sublimator replied:
just saw this

sublimator replied:
addressed

armfazh commented on src/groupSjcl.ts:
```diff
@@ -482,7 +473,9 @@ class GroupSj implements Group {
     static readonly Scalar = ScalarSj
     static readonly ID = GroupIDs
-    static readonly getID = getGroupID
+    static for(gid: GroupID) {
```
see my comment above about the `for` keyword.

armfazh commented on src/groupTypes.ts:
```diff
@@ -107,9 +94,8 @@ export interface Group extends SerializationHelpers {
 
 export interface GroupCons {
-    new (id: GroupID): Group
+    for(id: GroupID): Group
```
see comment above about using `for` keyword

armfazh commented:
@sublimator please rebase and squash your commits, so it can be merged.

sublimator commented:
I got conflicts when I tried to rebase, so I just merged origin/main, so it should squash down from the Github UI easily.

Only merge commits allowed?

sublimator commented:
Ok, I created a new commit on top of main from the contents of the old commits and force pushed.

armfazh commented:
Thank for your contributions @sublimator

sublimator commented:
Wooo! Thanks

On Sat, 2 Sep 2023 at 7:04 AM Armando Faz ***@***.***> wrote:

> Thank for your contributions @sublimator <https://github.com/sublimator>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cloudflare/voprf-ts/pull/38#issuecomment-1703535352>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEAHG64XUJ476C6I6PP34LXYJZ2JANCNFSM6AAAAAAZNF5HXI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
